binlog_driver/reader/offset.go

```diff
@@ -0,0 +1,167 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package reader
+
+import (
+	"github.com/Shopify/sarama"
+	"github.com/ngaut/log"
+	pb "github.com/pingcap/tidb-tools/binlog_proto/go-binlog"
+)
+
+// KafkaSeeker implements Kafka Seeker
+type KafkaSeeker struct {
+	consumer sarama.Consumer
+	client   sarama.Client
+}
+
+// NewKafkaSeeker return a instance of KafkaSeeker
+func NewKafkaSeeker(addr []string, config *sarama.Config) (*KafkaSeeker, error) {
+	client, err := sarama.NewClient(addr, config)
+	if err != nil {
+		return nil, err
+	}
+
+	consumer, err := sarama.NewConsumerFromClient(client)
+	if err != nil {
+		return nil, err
+	}
+
+	s := &KafkaSeeker{
+		client:   client,
+		consumer: consumer,
+	}
+
+	return s, nil
+}
+
+// Close release resource of KafkaSeeker
+func (ks *KafkaSeeker) Close() {
+	ks.consumer.Close()
+	ks.client.Close()
+}
+
+// Seek return the first offset which binlog CommitTs bigger than ts
+func (ks *KafkaSeeker) Seek(topic string, ts int64, partitions []int32) (offsets []int64, err error) {
+	if len(partitions) == 0 {
+		partitions, err = ks.consumer.Partitions(topic)
+		if err != nil {
+			log.Errorf("get partitions from topic %s error %v", topic, err)
+			return nil, err
+		}
+	}
+
+	offsets, err = ks.seekOffsets(topic, partitions, ts)
+	if err != nil {
+		log.Errorf("seek offsets error %v", err)
+		return
+	}
+	return
+}
+
+func (ks *KafkaSeeker) getTSFromMSG(msg *sarama.ConsumerMessage) (ts int64, err error) {
+	binlog := new(pb.Binlog)
+	err = binlog.Unmarshal(msg.Value)
+	if err != nil {
+		return
+	}
+
+	return binlog.CommitTs, nil
+}
+
+// seekOffsets returns all valid offsets in partitions
+func (ks *KafkaSeeker) seekOffsets(topic string, partitions []int32, pos int64) ([]int64, error) {
+	offsets := make([]int64, len(partitions))
+	for _, partition := range partitions {
+		start, err := ks.client.GetOffset(topic, partition, sarama.OffsetOldest)
+		if err != nil {
+			return nil, err
+		}
+
+		end, err := ks.client.GetOffset(topic, partition, sarama.OffsetNewest)
+		if err != nil {
+			return nil, err
+		}
+
+		offset, err := ks.seekOffset(topic, partition, start, end-1, pos)
+		if err != nil {
+			return nil, err
+		}
+		offsets[partition] = offset
+	}
+
+	return offsets, nil
+}
+
+func (ks *KafkaSeeker) seekOffset(topic string, partition int32, start int64, end int64, ts int64) (offset int64, err error) {
+
+	startTS, err := ks.getTSAtOffset(topic, partition, start)
+	if err != nil {
+		return
+	}
+
+	if ts < startTS {
+		log.Warnf("given ts %v is smaller than oldest message's ts %v, some binlogs may lose", ts, startTS)
```

Comment:
Put it at L118, and check `if ts == startTS return start+1`

Reply:
done